## <div align="center"><b> Проект по SQL </b></div>

### Цели исследования

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.  
Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша цель — <b>проанализировать базу данных.</b> В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут <b>сформулировать ценностное предложение для нового продукта.</b>

### Описание данных

<b>Таблица books</b>  
Содержит данные о книгах:  
- book_id — идентификатор книги;  
- author_id — идентификатор автора;  
- title — название книги;  
- num_pages — количество страниц;  
- publication_date — дата публикации книги;  
- publisher_id — идентификатор издателя.  

<b>Таблица authors</b>  
Содержит данные об авторах:  
- author_id — идентификатор автора;  
- author — имя автора.  

<b>Таблица publishers</b>  
Содержит данные об издательствах:  
- publisher_id — идентификатор издательства;  
- publisher — название издательства;  

<b>Таблица ratings</b>  
Содержит данные о пользовательских оценках книг:  
- rating_id — идентификатор оценки;  
- book_id — идентификатор книги;  
- username — имя пользователя, оставившего оценку;  
- rating — оценка книги.  

<b>Таблица reviews</b>  
Содержит данные о пользовательских обзорах:  
- review_id — идентификатор обзора;  
book_id — идентификатор книги;  
username — имя автора обзора;  
text — текст обзора.

### Задания

- Посчитать, сколько книг вышло после 1 января 2000 года;  
- Для каждой книги посчитать количество обзоров и среднюю оценку;  
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;  
- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;  
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Импортируем библиотеки

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine


### Создаём коннекцию к базам и выводим первые строки

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
query = '''
            SELECT * FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)

In [4]:
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
# получаем сырые данные
query = '''
            SELECT * FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)

In [6]:
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
# получаем сырые данные
query = '''
            SELECT * FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)

In [8]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
# получаем сырые данные
query = '''
            SELECT * FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)

In [10]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
# получаем сырые данные
query = '''
            SELECT * FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)

In [12]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
# КОД РЕВЬЮЕРА
# какие ещё таблицы доступны в базе
display(pd.io.sql.read_sql('''

SELECT *
FROM pg_catalog.pg_tables
WHERE schemaname != 'pg_catalog' AND
      schemaname != 'information_schema';

''', con = engine))


# смотрим на типы столбцов в интересующих таблицах
display(pd.io.sql.read_sql('''
SELECT 
    table_name, 
    column_name, 
    data_type, 
    is_nullable
FROM INFORMATION_SCHEMA.COLUMNS 
WHERE table_name IN ('books', 'authors', 'publishers', 'ratings', 'reviews');
''', con = engine))

schemaname           tablename       tableowner tablespace  hasindexes  \
0     public              orders  praktikum_admin       None        True   
1     public              visits  praktikum_admin       None        True   
2     public  advertisment_costs  praktikum_admin       None        True   
3     public             authors  praktikum_admin       None        True   
4     public          publishers  praktikum_admin       None        True   
5     public             reviews  praktikum_admin       None        True   
6     public             ratings  praktikum_admin       None        True   
7     public               books  praktikum_admin       None        True   

   hasrules  hastriggers  rowsecurity  
0     False        False        False  
1     False        False        False  
2     False        False        False  
3     False         True        False  
4     False         True        False  
5     False         True        False  
6     False         True        False  
7     False         True        False

table_name       column_name data_type is_nullable
0      authors         author_id   integer          NO
1      authors            author      text         YES
2   publishers      publisher_id   integer          NO
3   publishers         publisher      text         YES
4      reviews         review_id   integer          NO
5      reviews           book_id   integer         YES
6      reviews          username      text         YES
7      reviews              text      text         YES
8      ratings         rating_id   integer          NO
9      ratings           book_id   integer         YES
10     ratings          username      text         YES
11     ratings            rating   integer         YES
12       books           book_id   integer          NO
13       books         author_id   integer         YES
14       books             title      text         YES
15       books         num_pages   integer         YES
16       books  publication_date      date         YES
17       books      publisher_id   integer         YES

### Задание 1 - посчитать, сколько книг вышло после 1 января 2000 года.

In [14]:
# функция для запросов к данным
def select(sql):
    return pd.read_sql(sql, con=engine)

In [15]:
task1 = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1'; 
'''
select(task1)

count
0    819

<b>Вывод:</b>  
После 1 января 2000 года вышло 819 книг.

### Задание 2 - для каждой книги посчитать количество обзоров и среднюю оценку.

In [16]:
task2 = '''                              
SELECT
    b.book_id AS books_id,
    b.title AS books_title,
    COUNT(DISTINCT rw.review_id) AS counts_reviews,
    AVG(rt.rating) AS avg_ratings

FROM
    books AS b
    INNER JOIN ratings AS rt ON rt.book_id = b.book_id
    LEFT JOIN reviews AS rw ON rw.book_id = b.book_id
                    

GROUP BY
    b.title,
    b.book_id

ORDER BY
    counts_reviews DESC
LIMIT 20                        
; ''' 
select(task2).round(2)

books_id                                        books_title  \
0        948                            Twilight (Twilight  #1)   
1        627                                      The Alchemist   
2        497                          Outlander (Outlander  #1)   
3        750                The Hobbit  or There and Back Again   
4        207                                    Eat  Pray  Love   
5        656                                     The Book Thief   
6        695  The Curious Incident of the Dog in the Night-Time   
7        673                             The Catcher in the Rye   
8        733                          The Giver (The Giver  #1)   
9        734                                   The Glass Castle   
10       779  The Lightning Thief (Percy Jackson and the Oly...   
11       854                                           The Road   
12       963                                Water for Elephants   
13       696             The Da Vinci Code (Robert Langdon  #2)   
14       299  Harry Potter and the Chamber of Secrets (Harry...   
15       302  Harry Potter and the Prisoner of Azkaban (Harr...   
16        15     A Clash of Kings  (A Song of Ice and Fire  #2)   
17       232                Extremely Loud and Incredibly Close   
18       267  Freakonomics: A Rogue Economist Explores the H...   
19       194                                            Dracula   

    counts_reviews  avg_ratings  
0                7         3.66  
1                6         3.79  
2                6         4.12  
3                6         4.12  
4                6         3.40  
5                6         4.26  
6                6         4.08  
7                6         3.83  
8                6         3.75  
9                6         4.21  
10               6         4.08  
11               6         3.77  
12               6         3.98  
13               6         3.83  
14               6         4.29  
15               6         4.41  
16               5         4.39  
17               5         3.42  
18               5         3.91  
19               5         3.30

Наибольшее количество обзоров получила книга под названиями Twilight (Twilight #1) со средней оценкой 3,66.                                                                    

### Задание 3 - определить издательство, которое выпустило наибольшее число книг толще 50 страниц.

In [17]:
task3 = ''' 
SELECT
    table1.publisher_id,
    table1.name_publisher,
    table1.count_books
FROM
    (SELECT 
        b.publisher_id AS publisher_id,
        COUNT(b.book_id) AS count_books,
        p.publisher AS name_publisher 
    FROM 
        books AS b
        LEFT JOIN publishers AS p ON b.publisher_id=p.publisher_id

    WHERE 
        num_pages>50
    GROUP BY
        b.publisher_id,
        p.publisher
    ORDER BY 
        count_books DESC) AS table1
; ''' 
select(task3)

publisher_id                         name_publisher  count_books
0             212                          Penguin Books           42
1             309                                Vintage           31
2             116               Grand Central Publishing           25
3             217                       Penguin Classics           24
4              33                       Ballantine Books           19
..            ...                                    ...          ...
329           148          Harvard Business Review Press            1
330           302               Tyndale House Publishers            1
331            66                        Corgi Childrens            1
332           230  Pocket Books / Simon & Schuster  Inc.            1
333           199                          Orchard Books            1

[334 rows x 3 columns]

<b>Вывод:</b>  
Издательство 'Penguin Books'  выпустило наибольшее число книг (42 книги) толщиной более 50 страниц.

### Задание 4 - определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.

In [18]:
task4 = ''' 
SELECT 
    avg_aut_rat.author,
    AVG(avg_aut_rat.avg_rating) AS avg_aut_rating
FROM
        (SELECT 
            a.author AS author,
            AVG(r.rating) AS avg_rating
        FROM books AS b
            LEFT JOIN authors AS a ON b.author_id = a.author_id
            LEFT JOIN ratings AS r ON b.book_id = r.book_id
        GROUP BY 
             b.book_id,
             a.author_id
        HAVING
            COUNT(r.rating_id) >=50 
        ) AS avg_aut_rat
GROUP BY    
    avg_aut_rat.author
ORDER BY
    avg_aut_rating DESC
    
; ''' 
select(task4)

author  avg_aut_rating
0                          J.K. Rowling/Mary GrandPré        4.283844
1                   Markus Zusak/Cao Xuân Việt Khương        4.264151
2                                      J.R.R. Tolkien        4.258446
3                                   Louisa May Alcott        4.192308
4                                        Rick Riordan        4.080645
5                                     William Golding        3.901408
6                                       J.D. Salinger        3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce        3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...        3.787879
9                                           Dan Brown        3.754540
10                                         Lois Lowry        3.750000
11            George Orwell/Boris Grabnar/Peter Škerl        3.729730
12                                    Stephenie Meyer        3.662500
13                                     John Steinbeck        3.622951

<b>Вывод:</b>  
Если учитывать только книги с 50 и более оценками, то автором с самой высокой средней оценкой книг становится J.K. Rowling/Mary GrandPré

### Задание 5 - посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [19]:
task5 = ''' 

SELECT
    AVG(i.count_rev_id)
FROM
    (SELECT
        DISTINCT(username),
        COUNT(review_id) AS count_rev_id
    FROM
        reviews
    WHERE 
        username IN (
                SELECT 
                    DISTINCT(username)
                FROM 
                    ratings
                GROUP BY 
                    username
                HAVING 
                    COUNT(rating_id) > 50
                    )
    GROUP BY 
        username
    ) AS i
; ''' 
select(task5)

avg
0  24.333333

<b>Вывод:</b>  
Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24,33.